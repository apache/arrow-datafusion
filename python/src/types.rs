// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use datafusion::arrow::datatypes::DataType;
use pyo3::{FromPyObject, PyAny, PyResult};

use crate::pyarrow::PyArrowConvert;

/// utility struct to convert PyObj to native DataType
#[derive(Debug, Clone)]
pub struct PyDataType {
    pub data_type: DataType,
}

impl<'source> FromPyObject<'source> for PyDataType {
    fn extract(ob: &'source PyAny) -> PyResult<Self> {
        let dtype = DataType::from_pyarrow(ob)?;
        Ok(PyDataType { data_type: dtype })
    }
}
