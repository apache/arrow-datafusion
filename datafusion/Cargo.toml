# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

[package]
authors = ["Apache Arrow <dev@arrow.apache.org>"]
description = "DataFusion is an in-memory query engine that uses Apache Arrow as the memory model"
edition = "2021"
homepage = "https://github.com/apache/arrow-datafusion"
include = [
  "benches/*.rs",
  "src/**/*.rs",
  "Cargo.toml",
]
keywords = ["arrow", "query", "sql"]
license = "Apache-2.0"
name = "datafusion"
readme = "../README.md"
repository = "https://github.com/apache/arrow-datafusion"
rust-version = "1.58"
version = "6.0.0"

[lib]
name = "datafusion"
path = "src/lib.rs"

[features]
crypto_expressions = ["md-5", "sha2", "blake2", "blake3"]
default = ["crypto_expressions", "regex_expressions", "unicode_expressions"]
pyarrow = ["pyo3", "arrow/pyarrow"]
regex_expressions = ["regex"]
simd = ["arrow/simd"]
unicode_expressions = ["unicode-segmentation"]
# Used for testing ONLY: causes all values to hash to the same value (test for collisions)
force_hash_collisions = []
# Used to enable the avro format
avro = ["avro-rs", "num-traits"]

[dependencies]
ahash = {version = "0.7", default-features = false}
arrow = {version = "8.0.0", features = ["prettyprint"]}
async-trait = "0.1.41"
avro-rs = {version = "0.13", features = ["snappy"], optional = true}
avro-rs = {version = "0.13", features = ["snappy"], optional = true}
blake2 = {version = "^0.9.2", optional = true}
blake2 = {version = "^0.10.2", optional = true}
blake3 = {version = "1.0", optional = true}
blake3 = {version = "1.0", optional = true}
chrono = {version = "0.4", default-features = false}
chrono = {version = "0.4", default-features = false}
futures = "0.3"
hashbrown = {version = "0.12", features = ["raw"]}
hashbrown = {version = "0.11", features = ["raw"]}
lazy_static = {version = "^1.4.0"}
log = "^0.4"
md-5 = {version = "^0.10.0", optional = true}
num = "0.4"
num-traits = {version = "0.2", optional = true}
num_cpus = "1.13.0"
ordered-float = "2.0"
parquet = {version = "8.0.0", features = ["arrow"]}
parquet = {version = "7.0.0", features = ["arrow"]}
paste = "^1.0"
paste = "^1.0"
pin-project-lite = "^0.2.7"
pyo3 = {version = "0.14", optional = true}
pyo3 = {version = "0.15", optional = true}
rand = "0.8"
sha2 = {version = "^0.10.1", optional = true}
sqlparser = "0.13"
tempfile = "3"
tokio = {version = "1.0", features = ["macros", "rt", "rt-multi-thread", "sync", "fs"]}
tokio-stream = "0.1"
unicode-segmentation = {version = "^1.7.1", optional = true}

[dev-dependencies]
criterion = "0.3"
doc-comment = "0.3"

[[bench]]
harness = false
name = "aggregate_query_sql"

[[bench]]
harness = false
name = "sort_limit_query_sql"

[[bench]]
harness = false
name = "math_query_sql"

[[bench]]
harness = false
name = "filter_query_sql"

[[bench]]
harness = false
name = "window_query_sql"

[[bench]]
harness = false
name = "scalar"

[[bench]]
harness = false
name = "physical_plan"
