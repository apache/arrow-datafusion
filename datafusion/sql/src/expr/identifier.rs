// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use crate::planner::{ContextProvider, PlannerContext, SqlToRel};
use crate::utils::normalize_ident;
use datafusion_common::{
    Column, DFSchema, DataFusionError, Result, ScalarValue, TableReference,
};
use datafusion_expr::{Case, Expr, GetIndexedField};
use sqlparser::ast::{Expr as SQLExpr, Ident};

impl<'a, S: ContextProvider> SqlToRel<'a, S> {
    pub(super) fn sql_identifier_to_expr(&self, id: Ident) -> Result<Expr> {
        if id.value.starts_with('@') {
            // TODO: figure out if ScalarVariables should be insensitive.
            let var_names = vec![id.value];
            let ty = self
                .schema_provider
                .get_variable_type(&var_names)
                .ok_or_else(|| {
                    DataFusionError::Execution(format!(
                        "variable {var_names:?} has no type information"
                    ))
                })?;
            Ok(Expr::ScalarVariable(ty, var_names))
        } else {
            // Don't use `col()` here because it will try to
            // interpret names with '.' as if they were
            // compound identifiers, but this is not a compound
            // identifier. (e.g. it is "foo.bar" not foo.bar)

            Ok(Expr::Column(Column {
                relation: None,
                name: normalize_ident(id),
            }))
        }
    }

    // (relation, column name)
    fn form_identifier(idents: &[String]) -> Result<(Option<TableReference>, &String)> {
        match idents.len() {
            1 => Ok((None, &idents[0])),
            2 => Ok((
                Some(TableReference::Bare {
                    table: (&idents[0]).into(),
                }),
                &idents[1],
            )),
            3 => Ok((
                Some(TableReference::Partial {
                    schema: (&idents[0]).into(),
                    table: (&idents[1]).into(),
                }),
                &idents[2],
            )),
            4 => Ok((
                Some(TableReference::Full {
                    catalog: (&idents[0]).into(),
                    schema: (&idents[1]).into(),
                    table: (&idents[2]).into(),
                }),
                &idents[3],
            )),
            _ => Err(DataFusionError::Internal(format!(
                "Incorrect number of identifiers: {}",
                idents.len()
            ))),
        }
    }

    pub(super) fn sql_compound_identifier_to_expr(
        &self,
        ids: Vec<Ident>,
        schema: &DFSchema,
    ) -> Result<Expr> {
        if ids.len() < 2 {
            return Err(DataFusionError::Internal(format!(
                "Not a compound identifier: {ids:?}"
            )));
        }

        if ids[0].value.starts_with('@') {
            let var_names: Vec<_> = ids.into_iter().map(normalize_ident).collect();
            let ty = self
                .schema_provider
                .get_variable_type(&var_names)
                .ok_or_else(|| {
                    DataFusionError::Execution(format!(
                        "variable {var_names:?} has no type information"
                    ))
                })?;
            Ok(Expr::ScalarVariable(ty, var_names))
        } else {
            let ids = ids
                .into_iter()
                .map(|id| {
                    if self.options.enable_ident_normalization {
                        normalize_ident(id)
                    } else {
                        id.value
                    }
                })
                .collect::<Vec<_>>();

            // Possibilities we search with, in order from top to bottom for each len:
            //
            // len = 2:
            // 1. (table.column)
            // 2. (column).nested
            //
            // len = 3:
            // 1. (schema.table.column)
            // 2. (table.column).nested
            // 3. (column).nested1.nested2
            //
            // len = 4:
            // 1. (catalog.schema.table.column)
            // 2. (schema.table.column).nested1
            // 3. (table.column).nested1.nested2
            // 4. (column).nested1.nested2.nested3
            //
            // len = 5:
            // 1. (catalog.schema.table.column).nested
            // 2. (schema.table.column).nested1.nested2
            // 3. (table.column).nested1.nested2.nested3
            // 4. (column).nested1.nested2.nested3.nested4
            //
            // len > 5:
            // 1. (catalog.schema.table.column).nested[.nestedN]+
            // 2. (schema.table.column).nested1.nested2[.nestedN]+
            // 3. (table.column).nested1.nested2.nested3[.nestedN]+
            // 4. (column).nested1.nested2.nested3.nested4[.nestedN]+
            //
            // Currently not supporting more than one nested level
            // Though ideally once that support is in place, this code should work with it

            // TODO: remove when can support multiple nested identifiers
            if ids.len() > 5 {
                return Err(DataFusionError::Internal(format!(
                    "Unsupported compound identifier: {ids:?}"
                )));
            }

            // take at most 4 identifiers to form a Column to search with
            // - 1 for the column name
            // - 0 to 3 for the TableReference
            let bound = ids.len().min(4);
            // search from most specific to least specific
            let search_result = (0..bound).rev().find_map(|i| {
                let nested_names_index = i + 1;
                let s = &ids[0..nested_names_index];
                let (relation, column_name) = Self::form_identifier(s).unwrap();
                let field = schema.field_with_name(relation.as_ref(), column_name).ok();
                field.map(|f| (f, nested_names_index))
            });

            match search_result {
                // found matching field with spare identifier(s) for nested field(s) in structure
                Some((field, index)) if index < ids.len() => {
                    // TODO: remove when can support multiple nested identifiers
                    if index < ids.len() - 1 {
                        return Err(DataFusionError::Internal(format!(
                            "Nested identifiers not yet supported for column {}",
                            field.qualified_column().quoted_flat_name()
                        )));
                    }
                    let nested_name = ids[index].to_string();
                    Ok(Expr::GetIndexedField(GetIndexedField::new(
                        Box::new(Expr::Column(field.qualified_column())),
                        ScalarValue::Utf8(Some(nested_name)),
                    )))
                }
                // found matching field with no spare identifier(s)
                Some((field, _index)) => Ok(Expr::Column(field.qualified_column())),
                // found no matching field, will return a default
                None => {
                    // return default where use all identifiers to not have a nested field
                    // this len check is because at 5 identifiers will have to have a nested field
                    if ids.len() == 5 {
                        Err(DataFusionError::Internal(format!(
                            "Unsupported compound identifier: {ids:?}"
                        )))
                    } else {
                        let s = &ids[0..ids.len()];
                        let (relation, column_name) = Self::form_identifier(s).unwrap();
                        let relation = relation.map(|r| r.to_owned_reference());
                        Ok(Expr::Column(Column::new(relation, column_name)))
                    }
                }
            }
        }
    }

    pub(super) fn sql_case_identifier_to_expr(
        &self,
        operand: Option<Box<SQLExpr>>,
        conditions: Vec<SQLExpr>,
        results: Vec<SQLExpr>,
        else_result: Option<Box<SQLExpr>>,
        schema: &DFSchema,
        planner_context: &mut PlannerContext,
    ) -> Result<Expr> {
        let expr = if let Some(e) = operand {
            Some(Box::new(self.sql_expr_to_logical_expr(
                *e,
                schema,
                planner_context,
            )?))
        } else {
            None
        };
        let when_expr = conditions
            .into_iter()
            .map(|e| self.sql_expr_to_logical_expr(e, schema, planner_context))
            .collect::<Result<Vec<_>>>()?;
        let then_expr = results
            .into_iter()
            .map(|e| self.sql_expr_to_logical_expr(e, schema, planner_context))
            .collect::<Result<Vec<_>>>()?;
        let else_expr = if let Some(e) = else_result {
            Some(Box::new(self.sql_expr_to_logical_expr(
                *e,
                schema,
                planner_context,
            )?))
        } else {
            None
        };

        Ok(Expr::Case(Case::new(
            expr,
            when_expr
                .iter()
                .zip(then_expr.iter())
                .map(|(w, t)| (Box::new(w.to_owned()), Box::new(t.to_owned())))
                .collect(),
            else_expr,
        )))
    }
}
