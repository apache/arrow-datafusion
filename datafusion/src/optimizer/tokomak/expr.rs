use crate::{
    error::DataFusionError,
    logical_plan::window_frames::WindowFrame,
    optimizer::tokomak::{ScalarUDFName, SortSpec, UDAFName, UDFName},
    physical_plan::{
        aggregates::AggregateFunction, functions::BuiltinScalarFunction,
        window_functions::WindowFunction,
    },
    scalar::ScalarValue,
};
use egg::*;

use super::datatype::TokomakDataType;
use super::scalar::TokomakScalar;
use std::convert::TryInto;
use std::str::FromStr;
define_language! {
pub enum TokomakExpr {
    "+" = Plus([Id; 2]),
    "-" = Minus([Id; 2]),
    "*" = Multiply([Id; 2]),
    "/" = Divide([Id; 2]),
    "%" = Modulus([Id; 2]),
    "not" = Not(Id),
    "or" = Or([Id; 2]),
    "and" = And([Id; 2]),
    "=" = Eq([Id; 2]),
    "<>" = NotEq([Id; 2]),
    "<" = Lt([Id; 2]),
    "<=" = LtEq([Id; 2]),
    ">" = Gt([Id; 2]),
    ">=" = GtEq([Id; 2]),
    "regex_match"=RegexMatch([Id;2]),
    "regex_imatch"=RegexIMatch([Id;2]),
    "regex_not_match"=RegexNotMatch([Id;2]),
    "regex_not_imatch"=RegexNotIMatch([Id;2]),


    "is_not_null" = IsNotNull(Id),
    "is_null" = IsNull(Id),
    "negative" = Negative(Id),
    "between" = Between([Id; 3]),
    "between_inverted" = BetweenInverted([Id; 3]),
    "like" = Like([Id; 2]),
    "not_like" = NotLike([Id; 2]),
    "in_list" = InList([Id; 2]),
    "not_in_list" = NotInList([Id; 2]),
    "list" = List(Vec<Id>),

    //ScalarValue types

    Type(TokomakDataType),
    ScalarBuiltin(BuiltinScalarFunction),
    AggregateBuiltin(AggregateFunction),
    WindowBuiltin(WindowFunction),
    Scalar(TokomakScalar),

    //THe fist expression for all of the function call types must be the corresponding function type
    //For UDFs this is a string, which is looked up in the ExecutionProps
    //The last expression must be a List and is the arguments for the function.
    "call" = ScalarBuiltinCall([Id; 2]),
    "call_udf"=ScalarUDFCall([Id; 2]),
    "call_agg" = AggregateBuiltinCall([Id; 2]),
    "call_agg_distinct"=AggregateBuiltinDistinctCall([Id;2]),
    "call_udaf" = AggregateUDFCall([Id; 2]),
    //For window fuctions index 1 is the window partition
    //index 2 is the window order
    "call_win" = WindowBuiltinCallUnframed([Id;4 ]),
    //For a framed window function index 3 is the frame parameters
    "call_win_framed" = WindowBuiltinCallFramed([Id;5 ]),
    //Last Id of the Sort node MUST be a SortSpec
    "sort" = Sort([Id;2]),
    SortSpec(SortSpec),
    ScalarUDF(ScalarUDFName),
    AggregateUDF(UDAFName),
    Column(Symbol),
    WindowFrame(WindowFrame),

    // cast id as expr. Type is encoded as symbol
    "cast" = Cast([Id; 2]),
    "try_cast" = TryCast([Id;2]),
}
}

impl TokomakExpr {
    pub(crate) fn can_convert_to_scalar_value(&self) -> bool {
        matches!(self, TokomakExpr::Scalar(_))
    }
}

impl TryInto<ScalarValue> for &TokomakExpr {
    type Error = DataFusionError;

    fn try_into(self) -> Result<ScalarValue, Self::Error> {
        match self {
            TokomakExpr::Scalar(s) => Ok(s.clone().into()),
            e => Err(DataFusionError::Internal(format!(
                "Could not convert {:?} to scalar",
                e
            ))),
        }
    }
}

impl From<ScalarValue> for TokomakExpr {
    fn from(v: ScalarValue) -> Self {
        TokomakExpr::Scalar(v.into())
    }
}
impl FromStr for WindowFrame {
    type Err = DataFusionError;
    #[allow(unused_variables)]
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Err(DataFusionError::NotImplemented(
            "WindowFrame's aren't parsed yet".to_string(),
        ))
    }
}
