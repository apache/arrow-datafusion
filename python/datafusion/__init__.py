from .internals import PyDataFrame as DataFrame
from .internals import PyExecutionContext as ExecutionContext
from .internals import PyExpr as Expr
from .internals import functions
