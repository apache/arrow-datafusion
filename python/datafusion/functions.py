from .internals import functions


def __getattr__(name):
    return getattr(functions, name)
