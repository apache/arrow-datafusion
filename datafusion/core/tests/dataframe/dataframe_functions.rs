// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use arrow::datatypes::{DataType, Field, Schema};
use arrow::{
    array::{Int32Array, StringArray},
    record_batch::RecordBatch,
};
use arrow_array::types::Int32Type;
use arrow_array::ListArray;
use arrow_schema::SchemaRef;
use std::sync::Arc;

use datafusion::error::Result;

use datafusion::prelude::*;

use datafusion::assert_batches_eq;
use datafusion_common::{assert_snapshot, DFSchema, ScalarValue};
use datafusion_expr::expr::Alias;
use datafusion_expr::ExprSchemable;

fn test_schema() -> SchemaRef {
    Arc::new(Schema::new(vec![
        Field::new("a", DataType::Utf8, false),
        Field::new("b", DataType::Int32, false),
        Field::new("l", DataType::new_list(DataType::Int32, true), true),
    ]))
}

async fn create_test_table() -> Result<DataFrame> {
    let schema = test_schema();

    // define data.
    let batch = RecordBatch::try_new(
        schema,
        vec![
            Arc::new(StringArray::from(vec![
                "abcDEF",
                "abc123",
                "CBAdef",
                "123AbcDef",
            ])),
            Arc::new(Int32Array::from(vec![1, 10, 10, 100])),
            Arc::new(ListArray::from_iter_primitive::<Int32Type, _, _>(vec![
                Some(vec![Some(0), Some(1), Some(2)]),
                None,
                Some(vec![Some(3), None, Some(5)]),
                Some(vec![Some(6), Some(7)]),
            ])),
        ],
    )?;

    let ctx = SessionContext::new();

    ctx.register_batch("test", batch)?;

    ctx.table("test").await
}

/// Executes an expression on the test dataframe as a select.
/// Compares formatted output of a record batch with an expected
/// vector of strings, using the assert_batch_eq! macro
macro_rules! assert_fn_batches {
    ($EXPR:expr) => {
        assert_fn_batches!($EXPR, 10)
    };
    ($EXPR:expr, $LIMIT: expr) => {
        let df = create_test_table().await?;
        let df = df.select(vec![$EXPR])?.limit(0, Some($LIMIT))?;
        let batches = df.collect().await?;

        // Once result is updated, run command below for snapshot update
        // `cargo insta test --accept -p datafusion --test core_integration -- dataframe::dataframe_functions`
        datafusion_common::assert_snapshot!(&batches);
    };
}

#[tokio::test]
async fn test_fn_ascii() -> Result<()> {
    let expr = ascii(col("a"));

    assert_fn_batches!(expr, 1);

    Ok(())
}

#[tokio::test]
async fn test_fn_bit_length() -> Result<()> {
    let expr = bit_length(col("a"));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_btrim() -> Result<()> {
    let expr = btrim(vec![lit("      a b c             ")]);

    assert_fn_batches!(expr, 1);

    Ok(())
}

#[tokio::test]
async fn test_fn_btrim_with_chars() -> Result<()> {
    let expr = btrim(vec![col("a"), lit("ab")]);

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_nullif() -> Result<()> {
    let expr = nullif(col("a"), lit("abcDEF"));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_arrow_cast() -> Result<()> {
    let expr = arrow_typeof(arrow_cast(col("b"), lit("Float64")));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_nvl() -> Result<()> {
    let lit_null = lit(ScalarValue::Utf8(None));
    // nvl(CASE WHEN a = 'abcDEF' THEN NULL ELSE a END, 'TURNED_NULL')
    let expr = nvl(
        when(col("a").eq(lit("abcDEF")), lit_null)
            .otherwise(col("a"))
            .unwrap(),
        lit("TURNED_NULL"),
    )
    .alias("nvl_expr");

    assert_fn_batches!(expr);

    Ok(())
}
#[tokio::test]
async fn test_nvl2() -> Result<()> {
    let lit_null = lit(ScalarValue::Utf8(None));
    // nvl2(CASE WHEN a = 'abcDEF' THEN NULL ELSE a END, 'NON_NUll', 'TURNED_NULL')
    let expr = nvl2(
        when(col("a").eq(lit("abcDEF")), lit_null)
            .otherwise(col("a"))
            .unwrap(),
        lit("NON_NULL"),
        lit("TURNED_NULL"),
    )
    .alias("nvl2_expr");

    assert_fn_batches!(expr);

    Ok(())
}
#[tokio::test]
async fn test_fn_arrow_typeof() -> Result<()> {
    let expr = arrow_typeof(col("l"));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_struct() -> Result<()> {
    let expr = r#struct(vec![col("a"), col("b")]);

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_named_struct() -> Result<()> {
    let expr = named_struct(vec![lit("column_a"), col("a"), lit("column_b"), col("b")]);

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_coalesce() -> Result<()> {
    let expr = coalesce(vec![lit(ScalarValue::Utf8(None)), lit("ab")]);

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_approx_median() -> Result<()> {
    let expr = approx_median(col("b"));

    let expected = [
        "+-----------------------+",
        "| APPROX_MEDIAN(test.b) |",
        "+-----------------------+",
        "| 10                    |",
        "+-----------------------+",
    ];

    let df = create_test_table().await?;
    let batches = df.aggregate(vec![], vec![expr]).unwrap().collect().await?;

    assert_batches_eq!(expected, &batches);

    Ok(())
}

#[tokio::test]
async fn test_fn_approx_percentile_cont() -> Result<()> {
    let expr = approx_percentile_cont(col("b"), lit(0.5));

    let df = create_test_table().await?;
    let batches = df.aggregate(vec![], vec![expr]).unwrap().collect().await?;

    assert_snapshot!(&batches);

    // the arg2 parameter is a complex expr, but it can be evaluated to the literal value
    let alias_expr = Expr::Alias(Alias::new(
        cast(lit(0.5), DataType::Float32),
        None::<&str>,
        "arg_2".to_string(),
    ));
    let expr = approx_percentile_cont(col("b"), alias_expr);
    let df = create_test_table().await?;
    let batches = df.aggregate(vec![], vec![expr]).unwrap().collect().await?;

    assert_snapshot!(&batches);

    Ok(())
}

#[tokio::test]
#[cfg(feature = "unicode_expressions")]
async fn test_fn_character_length() -> Result<()> {
    let expr = character_length(col("a"));
    assert_fn_batches!(expr);
    Ok(())
}

#[tokio::test]
async fn test_fn_chr() -> Result<()> {
    let expr = chr(lit(128175));

    assert_fn_batches!(expr, 1);

    Ok(())
}

#[tokio::test]
async fn test_fn_initcap() -> Result<()> {
    let expr = initcap(col("a"));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
#[cfg(feature = "unicode_expressions")]
async fn test_fn_left() -> Result<()> {
    let expr = left(col("a"), lit(3));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_lower() -> Result<()> {
    let expr = lower(col("a"));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
#[cfg(feature = "unicode_expressions")]
async fn test_fn_lpad() -> Result<()> {
    let expr = lpad(vec![col("a"), lit(10)]);

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
#[cfg(feature = "unicode_expressions")]
async fn test_fn_lpad_with_string() -> Result<()> {
    let expr = lpad(vec![col("a"), lit(10), lit("*")]);

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_ltrim() -> Result<()> {
    let expr = ltrim(vec![lit("      a b c             ")]);

    assert_fn_batches!(expr, 1);

    Ok(())
}

#[tokio::test]
async fn test_fn_ltrim_with_columns() -> Result<()> {
    let expr = ltrim(vec![col("a")]);

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
#[cfg(feature = "unicode_expressions")]
async fn test_fn_md5() -> Result<()> {
    let expr = md5(col("a"));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
#[cfg(feature = "unicode_expressions")]
async fn test_fn_regexp_like() -> Result<()> {
    let expr = regexp_like(col("a"), lit("[a-z]"));

    let _expected = [
        "+-----------------------------------+",
        "| regexp_like(test.a,Utf8(\"[a-z]\")) |",
        "+-----------------------------------+",
        "| true                              |",
        "| true                              |",
        "| true                              |",
        "| true                              |",
        "+-----------------------------------+",
    ];

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
#[cfg(feature = "unicode_expressions")]
async fn test_fn_regexp_match() -> Result<()> {
    let expr = regexp_match(col("a"), lit("[a-z]"));

    let _expected = [
        "+------------------------------------+",
        "| regexp_match(test.a,Utf8(\"[a-z]\")) |",
        "+------------------------------------+",
        "| [a]                                |",
        "| [a]                                |",
        "| [d]                                |",
        "| [b]                                |",
        "+------------------------------------+",
    ];

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
#[cfg(feature = "unicode_expressions")]
async fn test_fn_regexp_replace() -> Result<()> {
    let expr = regexp_replace(col("a"), lit("[a-z]"), lit("x"), lit("g"));

    let _expected = [
        "+----------------------------------------------------------+",
        "| regexp_replace(test.a,Utf8(\"[a-z]\"),Utf8(\"x\"),Utf8(\"g\")) |",
        "+----------------------------------------------------------+",
        "| xxxDEF                                                   |",
        "| xxx123                                                   |",
        "| CBAxxx                                                   |",
        "| 123AxxDxx                                                |",
        "+----------------------------------------------------------+",
    ];

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_replace() -> Result<()> {
    let expr = replace(col("a"), lit("abc"), lit("x"));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_repeat() -> Result<()> {
    let expr = repeat(col("a"), lit(2));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
#[cfg(feature = "unicode_expressions")]
async fn test_fn_reverse() -> Result<()> {
    let expr = reverse(col("a"));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
#[cfg(feature = "unicode_expressions")]
async fn test_fn_right() -> Result<()> {
    let expr = right(col("a"), lit(3));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
#[cfg(feature = "unicode_expressions")]
async fn test_fn_rpad() -> Result<()> {
    let expr = rpad(vec![col("a"), lit(11)]);

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
#[cfg(feature = "unicode_expressions")]
async fn test_fn_rpad_with_characters() -> Result<()> {
    let expr = rpad(vec![col("a"), lit(11), lit("x")]);

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
#[cfg(feature = "unicode_expressions")]
async fn test_fn_sha224() -> Result<()> {
    let expr = sha224(col("a"));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_split_part() -> Result<()> {
    let expr = split_part(col("a"), lit("b"), lit(1));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_starts_with() -> Result<()> {
    let expr = starts_with(col("a"), lit("abc"));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_ends_with() -> Result<()> {
    let expr = ends_with(col("a"), lit("DEF"));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
#[cfg(feature = "unicode_expressions")]
async fn test_fn_strpos() -> Result<()> {
    let expr = strpos(col("a"), lit("f"));
    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
#[cfg(feature = "unicode_expressions")]
async fn test_fn_substr() -> Result<()> {
    let expr = substr(col("a"), lit(2));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_cast() -> Result<()> {
    let expr = cast(col("b"), DataType::Float64);

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_to_hex() -> Result<()> {
    let expr = to_hex(col("b"));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
#[cfg(feature = "unicode_expressions")]
async fn test_fn_translate() -> Result<()> {
    let expr = translate(col("a"), lit("bc"), lit("xx"));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_upper() -> Result<()> {
    let expr = upper(col("a"));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_encode() -> Result<()> {
    let expr = encode(col("a"), lit("hex"));

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_decode() -> Result<()> {
    // Note that the decode function returns binary, and the default display of
    // binary is "hexadecimal" and therefore the output looks like decode did
    // nothing. So compare to a constant.
    let df_schema = DFSchema::try_from(test_schema().as_ref().clone())?;
    let expr = decode(encode(col("a"), lit("hex")), lit("hex"))
        // need to cast to utf8 otherwise the default display of binary array is hex
        // so it looks like nothing is done
        .cast_to(&DataType::Utf8, &df_schema)?;

    assert_fn_batches!(expr);

    Ok(())
}

#[tokio::test]
async fn test_fn_array_to_string() -> Result<()> {
    let expr = array_to_string(col("l"), lit("***"));

    assert_fn_batches!(expr);

    Ok(())
}
