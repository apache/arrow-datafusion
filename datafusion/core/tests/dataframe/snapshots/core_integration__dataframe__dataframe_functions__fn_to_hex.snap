---
source: datafusion/core/tests/dataframe/dataframe_functions.rs
expression: actual_lines
---
- +----------------+
- "| to_hex(test.b) |"
- +----------------+
- "| 1              |"
- "| a              |"
- "| a              |"
- "| 64             |"
- +----------------+
